### DISCUSSION TITLE
New auto-colouring of nodes

### DISCUSSION THREAD

**Discussion by PaulWilliamWoods** [freeplane/freeplane#424] [Help] [answered]
I like the new auto-colouring of nodes, such that the top one and its children is a different colour to the next and its children. I can't find how this has been implemented in the styles or format sections. How has this been done, please? Also, how do I now apply this style to old maps? (I've tried the 'copy map styles from...' and 'replace map styles by...' but neither worked.)

**Comment by macmarrum** [comment, marked as answer]
I believe you are referring to `Apply level styles` (right-click on mind map background for the menu)

![2022-06-03_15,28,17](https://user-images.githubusercontent.com/7197131/172007802-62884ffa-f7fb-4f1e-8895-4ba57755c0dc.png)

The colors usually come from the template used when creating the mind map. The "level styles" can be changed in `Format->Manage Styles->Edit styles`

**Reply by PaulWilliamWoods** [reply]
![image](https://user-images.githubusercontent.com/106873812/172010433-71763afc-b78e-40b3-9fad-c588eb0c23d4.png)
Perhaps you were referring to the 'automatic edge color' option in the above picture? The way my new maps behave suggests that this is set to 'on branch creation'. This seems to be what I'm referring to - I've never noticed it before. However, I'm still stuck about reverse-applying this to other maps, as I've tried using a map with this setting on, like the one in my picture, but it doesn't do the right thing to the map I'm trying to alter. It updates things like the width settings of nodes but not the colours.

**Reply by macmarrum** [reply]
It would be easier if I could open your mind map in my Freeplane.

You can replace each letter with 'x' before attaching it here, if you want to avoid exposing your content:

`File->Save map as`
`Edit->Find->Find and replace` 
Find `.` [x] Regular expressions
Replace `x`
[Replace all]

Please keep in mind that GitHub will allow only a zipped mind map to be attached.

**Reply by PaulWilliamWoods** [reply]
[Help with formatting.zip](https://github.com/freeplane/freeplane/files/8838174/Help.with.formatting.zip)
[New Mindmap.zip](https://github.com/freeplane/freeplane/files/8838177/New.Mindmap.zip)

**Reply by macmarrum** [reply]
I can see that `New Mindmap` has the option `Automatic edge color` set to `On branch creation`. This means that each time a new branch is created, it is assigned the next color from the list available at `Edit edge color` (in `Tool panel`).

Because `Help with formatting` already has branches, you could use `for branches` to color existing branches.

NB `Edit edge color` is only active when `Automatic edge color` is not disabled.

In `Help with formatting`, when I click `Edit edge color`, I see no colors defined, unlike in `New Mindmap`.

To have the same colors in `Help with formatting` as in `New Mindmap`, you need to add them one by one. 
Alternatively, you can open `New Mindmap.mm` in a text editor (notepad++, VS Code, ...), copy `edgeColorConfiguration=...` and paste to the same xml node in `Help with formatting.mm`.

```xml
<map version="freeplane 1.9.13">
<!--To view this file, download free mind mapping software Freeplane from https://www.freeplane.org -->
<node LOCALIZED_TEXT="new_mindmap" LOCALIZED_STYLE_REF="AutomaticLayout.level,0" FOLDED="false" ID="ID_696401721" CREATED="1610381621824" MODIFIED="1648901279394">
<hook NAME="AutomaticEdgeColor" COUNTER="7" RULE="ON_BRANCH_CREATION"/>
<hook NAME="MapStyle" background="#ffffff" zoom="1.464">
    <properties edgeColorConfiguration="#808080ff,#000000ff,#ff0033ff,#009933ff,#3333ffff,#ff6600ff,#cc00ccff,#ffbf00ff,#00ff99ff,#0099ffff,#996600ff,#000000ff,#cc0066ff,#33ff00ff,#ff9999ff,#0000ccff,#cccc00ff,#0099ccff,#006600ff,#ff00ccff,#00cc00ff,#0066ccff,#00ffffff" show_icon_for_attributes="true" associatedTemplateLocation="template:/My%20Default%20Format.mm" show_note_icons="true" fit_to_viewport="false"/>

<map_styles>
```

**Reply by PaulWilliamWoods** [reply]
Super - thank you for this. Your answer has comprehensively explained why my document didn't change when I selected 'for branches'. (It was missing the colours data.)

I presume that when I have done these edits but there's the odd errant node as in the picture here, that is some sort of bug?
![image](https://user-images.githubusercontent.com/106873812/172026443-f5abf708-b7eb-406b-a7d8-77281efc6a22.png)

**Reply by macmarrum** [reply]
I am glad it helped 😄 

Regarding your odd errant node, it probably has a an explicitly applied style or individual formatting applied. The actual mind map would help for me to investigate.

To better understand how formatting works, one needs to know that Freeplane uses a concept of layered Styles.

Here's the order of precedence in applying the different layers.

1. Selected style – if `Display selected nodes in bubbles` is off
2. Node individual formatting, including "automatic edge color: on branch creation"
3. Explicitly applied style
4. Node conditional styles
5. Map conditional styles
6. Automatic edge color: for branches / levels / columns
7. Automatic level style
8. Default style

`Default` is the base layer. Whatever is set on that layer will either shine through or be overwritten by the next layer put on top of `Default`.
`Automatic edge color: for branches / levels / columns` is above `Default`, therefore the edge color is shown. But when the node has an `explicitly applied style` (e.g. Floating node) or `node individual formatting` (in Tool panel) with a different edge color, it will overwrite the color of `Automatic edge color: for branches / levels / columns`.
The same goes for `selected style`, if used. If it has a (different) edge color set, it will overwrite `Automatic edge color: for branches / levels / columns`.

**Reply by PaulWilliamWoods** [reply]
Excellent - thank you again for this comprehensive response. That hierarchy will certainly help me with future diagnoses.

I wonder - is there a secondary documentation for Freeplane, other than the mind map in the application? (Apart from the actual source code, of course - I'm too rusty to make any fast use of it!)

**Reply by macmarrum** [reply]
😃 

Apart from `Help->Documentation` and `Help->Freeplane handbook`, there's https://freeplane.github.io/wiki/. It's a new wiki – work in progress.

The old wiki is available at https://www.freeplane.org/wiki/index.php/Main_Page. It's been recently migrated to https://freeplane.github.io/wiki/migrated-from-mediawiki/

**Reply by PaulWilliamWoods** [reply]
Thanks again! Most helpful.

**Reply by PaulWilliamWoods** [reply]
I have an addendum question: supposing I wanted to start the automatic branch colouring one level deeper, so that the auto-colours are applied not to the branches labelled 1,2,3 in the attached map but at the level labelled a,b,c - how does one go about this? The idea here is to use 1 as a collection label, like 'beginning' (2 would be 'middle' and 3 'end') and then each a,b,c, is a different item within that collection label. In my case, it's scenes in a novel, so it would be useful to have a,b,c differentiated by colour but it's also useful to see which subsection of the novel it's under at a glance (there will be more than 3 eventually).
[Later branch colours.zip](https://github.com/freeplane/freeplane/files/8872095/Later.branch.colours.zip)

**Reply by macmarrum** [reply]
`Automatic edge color: for branches` affects the entire mind map. To achieve what you described, I would use a script that colors the descendants of 1,2,3,... nodes (Automatic edge color: disabled). I would add a style `Edge-color-for-branches level 0` and mark 1,2,3,... with it.

In the script, the colors for subBranches are taken from the list at `Edit edge colors`.

The script needs to be run manually (on demand). Please let me know whether this is what you had in mind.
[PaulWilliamWood~Later branch colours~macmarrum.zip](https://github.com/freeplane/freeplane/files/8873063/PaulWilliamWood.Later.branch.colours.macmarrum.zip)


I realize the on-demand approach is inconvenient. It is possible to wrap the script in a NodeChangeListener that needs to be run only once per mind map (after it is opened) and then the script would fire automatically when nodes are added, deleted or moved.

```groovy
// @ExecutionModes({ON_SINGLE_NODE})

import org.freeplane.api.Node as FPN
import org.freeplane.features.edge.EdgeController
import org.freeplane.features.edge.mindmapmode.MEdgeController

final edgeColorForBranchesStyleName = 'Edge-color-for-branches level 0'

final mapModel = node.mindMap.delegate
final controller = EdgeController.controller as MEdgeController

c.find { it.style.name == edgeColorForBranchesStyleName }.each { FPN branchLevel0 ->
    branchLevel0.children.eachWithIndex { FPN branchLevel1, int i ->
        def colorCounter = i + 1
        def edgeColor = controller.getEdgeColor(mapModel, colorCounter)
        branchLevel1.findAll().each { level2plus -> level2plus.style.edge.color = edgeColor }
    }
}
```

**Reply by PaulWilliamWoods** [reply]
I'm getting an error executing the script, with explanation: null.

**Reply by PaulWilliamWoods** [reply]
Scrap that - I re-loaded and it worked fine, for whatever reason. Thank you, yes, that is the idea I had in mind regarding the colouring. You're correct thinking that I wanted it more automatic, so I'll consider your suggestion about the NodeChangeListener, but my coding ability isn't up to scratch to put this together and I don't have time right now to get it there! In the meantime, I will use summary nodes instead. It's significantly uglier and will no doubt cause me a lot of headaches when I'm re-arranging the map later, but it will do for now (especially having realised that summary nodes of summary nodes is possible and fits my purpose).

**Reply by macmarrum** [reply]
> my coding ability isn't up to scratch to put this together and I don't have time right now to get it there! In the meantime, I will use summary nodes instead.

Yep. I use summary nodes quite a lot.

I should have made it clear that I was happy to create a NodeChangeListener script, once the basic premise was confirmed. Anyway, I've realized that what is needed is actually a MapChangeListener, and that it is built into Freeplane → the Map Conditional Styles functionality. It could be used to run the script automatically, like in the attached mind map.
[PaulWilliamWoods~Later branch colours~macmarrum-v2.zip](https://github.com/freeplane/freeplane/files/8874028/PaulWilliamWoods.Later.branch.colours.macmarrum-v2.zip)

Just to explain, it uses a side effect of running Script Filter to find nodes for which a style is to be applied. The script, in addition to identifying the root node, runs a bunch of other code which identifies nodes and changes their edge color. The main action, as far as Map Conditional Styles is concerned, is to apply `Level 0` to the root node, though.

And here's the Script Filter contents (it's difficult to extract it from a mind map, once entered as part of a condition)
```groovy
if (!node.isRoot()) return false
//
import org.freeplane.api.Node as FPN
import org.freeplane.features.edge.EdgeController
import org.freeplane.features.edge.mindmapmode.MEdgeController

final edgeColorForBranchesStyleName = 'Edge-color-for-branches level 0'

final mapModel = node.mindMap.delegate
final controller = EdgeController.controller as MEdgeController

c.find { it.style.name == edgeColorForBranchesStyleName }.each { FPN branchLevel0 ->
    branchLevel0.children.eachWithIndex { FPN branchLevel1, int i ->
        def colorCounter = i + 1
        def edgeColor = controller.getEdgeColor(mapModel, colorCounter)
        branchLevel1.findAll().each { it.style.edge.color = edgeColor }
    }
}
return true
```

**Reply by PaulWilliamWoods** [reply]
Super! Thank you again. I'll try to incorporate this into the relevant map after the weekend deadline I'm working towards.

**Comment by euu2021** [comment]
Additional tip: As you are interested in coloring branches, you may also find interesting this script by Edo: https://sourceforge.net/p/freeplane/discussion/758437/thread/7e0c5f67ef/?limit=25

**Reply by PaulWilliamWoods** [reply]
Thanks; that's interesting - and if I get good at coding again, I might incorporate this into some artwork.
